### PR TITLE
api reward mapper: relegate correct validators

### PR DESCRIPTION
**Pull request by greywolve**
Establish the correct validator for rewards in the redelegate case.

### CODE REVIEW

**Comment by greywolve** [comment]
Need to add another test case or two

**Comment by greywolve** [comment]
I didn't really want to add a `client` to the mapper, but after considering all options I'd rather have this code in one place rather than adding a `client` to each `RewardProducer` on every worker.

I should probably make a `NewMapper` function rather too.

It's not great that api now depends on cosmosgrpc, but yeah... the RewardProducer flow in general seems to assume that we'd get enough information just from the txes themselves and not have to make additional calls, so there's not many good points to do an external call.

We'll probably need to add `ctx` to the RewardsProducer interface too, so it can be thread through. (For `MapTransactions`)

File: api/mapper/reward/reward.go [hunk 1]
```diff
@@ -350,6 +366,105 @@ func (m *Mapper) MsgBeginRedelegate(msg []byte, lg types.ABCIMessageLog) (rev *r
...
+					key := fmt.Sprintf("%s:%s", unc.ValidatorAddress, amt.Numeric.String())
+					if _, ok := usedValidators[key]; ok {
+						continue
+					}
+
+					ra := toDec(big.NewInt(0).SetBytes(rew.Amounts[0].Numeric), rew.Amounts[0].Exp)
+					ua := toDec(amt.Numeric, amt.Exp)
+
+					if rev.ValidatorSrc == unc.ValidatorAddress && ua.GTE(ra) {
+						ra := toDec(big.NewInt(0).SetBytes(rew.Amounts[0].Numeric), rew.Amounts[0].Exp)
```
Comment (tasinco) [inline api/mapper/reward/reward.go:414]: @greywolve this isn't needed b/c we calculate it on line 410?

File: api/mapper/reward/reward.go [hunk 2]
```diff
@@ -350,6 +366,105 @@ func (m *Mapper) MsgBeginRedelegate(msg []byte, lg types.ABCIMessageLog) (rev *r
...
+						ua := toDec(amt.Numeric, amt.Exp)
+						srcDelta = ra.Sub(ua)
+						srcExists = true
+						srckey = key
+						break
+					}
+
+					if rev.ValidatorDst == unc.ValidatorAddress && ua.GTE(ra) {
+						ra := toDec(big.NewInt(0).SetBytes(rew.Amounts[0].Numeric), rew.Amounts[0].Exp)
```
Comment (tasinco) [inline api/mapper/reward/reward.go:423]: @greywolve this isn't needed b/c we calculate it on line 410?

File: api/mapper/reward/reward.go [hunk 3]
```diff
@@ -350,6 +366,105 @@ func (m *Mapper) MsgBeginRedelegate(msg []byte, lg types.ABCIMessageLog) (rev *r
...
+						ua := toDec(amt.Numeric, amt.Exp)
+						dstDelta = ra.Sub(ua)
```
Comment (tasinco) [inline api/mapper/reward/reward.go:425]: @greywolve this should be ua.Sub(ra)

File: api/mapper/reward/reward.go [hunk 4]
```diff
@@ -350,6 +366,105 @@ func (m *Mapper) MsgBeginRedelegate(msg []byte, lg types.ABCIMessageLog) (rev *r
...
+						continue
+					}
+
+					ra := toDec(big.NewInt(0).SetBytes(rew.Amounts[0].Numeric), rew.Amounts[0].Exp)
+					ua := toDec(amt.Numeric, amt.Exp)
+
+					if rev.ValidatorSrc == unc.ValidatorAddress && ua.GTE(ra) {
+						ra := toDec(big.NewInt(0).SetBytes(rew.Amounts[0].Numeric), rew.Amounts[0].Exp)
+						ua := toDec(amt.Numeric, amt.Exp)
+						srcDelta = ra.Sub(ua)
```
Comment (tasinco) [inline api/mapper/reward/reward.go:416]: @greywolve this should be ua.Sub(ra)

File: api/mapper/reward/reward.go (hunk 3 above)
Comment (tasinco) [reply api/mapper/reward/reward.go]: ignore that....

File: api/mapper/reward/reward.go (hunk 4 above)
Comment (tasinco) [reply api/mapper/reward/reward.go]: ignore that...

**Comment by greywolve** [comment]
PR updated with all the various feedback :)

**Review by tasinco — APPROVED** [review]

**Pull request merged**